proxy/envoy/resources.go

```diff
@@ -211,40 +213,72 @@ type AccessLog struct {
 	Filter string `json:"filter,omitempty"`
 }
 
-// NetworkFilterConfig definition
-type NetworkFilterConfig struct {
-	CodecType         string       `json:"codec_type"`
-	StatPrefix        string       `json:"stat_prefix"`
-	GenerateRequestID bool         `json:"generate_request_id,omitempty"`
-	RouteConfig       *RouteConfig `json:"route_config"`
-	Filters           []Filter     `json:"filters"`
-	AccessLog         []AccessLog  `json:"access_log"`
-	Cluster           string       `json:"cluster,omitempty"`
+// HTTPFilterConfig definition
+type HTTPFilterConfig struct {
+	CodecType         string           `json:"codec_type"`
+	StatPrefix        string           `json:"stat_prefix"`
+	GenerateRequestID bool             `json:"generate_request_id,omitempty"`
+	RouteConfig       *HTTPRouteConfig `json:"route_config"`
+	Filters           []HTTPFilter     `json:"filters"`
+	AccessLog         []AccessLog      `json:"access_log"`
+}
+
+// TCPRoute definition
+type TCPRoute struct {
+	Cluster           string   `json:"cluster"`
+	DestinationIPList []string `json:"destination_ip_list,omitempty"`
+	DestinationPorts  []string `json:"destination_ports,omitempty"`
+	SourceIPList      []string `json:"source_ip_list,omitempty"`
+	SourcePorts       []string `json:"source_ports,omitempty"`
+}
+
+// TCPRouteConfigs provides routes by port
+type TCPRouteConfigs map[int]*TCPRouteConfig
+
+// TCPRouteConfig (or generalize as RouteConfig or L4RouteConfig for TCP/UDP?)
+type TCPRouteConfig struct {
+	Routes []TCPRoute `json:"routes"`
+}
+
+// EnsurePort creates a route config if necessary
+func (hosts TCPRouteConfigs) EnsurePort(port int) *TCPRouteConfig {
+	config, ok := hosts[port]
+	if !ok {
+		config = &TCPRouteConfig{}
+		hosts[port] = config
+	}
+	return config
+}
+
+// TCPProxyFilterConfig definition
+type TCPProxyFilterConfig struct {
+	StatPrefix  string         `json:"stat_prefix"`
+	RouteConfig TCPRouteConfig `json:"route_config"`
 }
 
 // NetworkFilter definition
 type NetworkFilter struct {
-	Type   string              `json:"type"`
-	Name   string              `json:"name"`
-	Config NetworkFilterConfig `json:"config"`
+	Type   string      `json:"type"`
+	Name   string      `json:"name"`
+	Config interface{} `json:"config"`
 }
 
 // Listener definition
 type Listener struct {
 	Port           int              `json:"port"`
 	Filters        []*NetworkFilter `json:"filters"`
-	BindToPort     bool             `json:"bind_to_port"`
+	BindToPort     bool             `json:"bind_to_port,omitempty"`
```

Comment:
bind_to_port = false will be omitted. please be careful with omitempty tags here.